File: 33-search-in-rotated-sorted-array/33-search-in-rotated-sorted-array.py
class Solution(object):
        
    def search(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: int
        """
        # edge cases:
        if len(nums) == 1:
            return 0 if nums[0] == target else -1
                
        left = 0
        right = len(nums)-1
        while right >= left: 
            mid = left + (right - left)//2

            if nums[mid] == target:
                return mid
            
            if nums[mid] > nums[right]:
                if target < nums[mid] and target >= nums[left]:
                    right = mid - 1
                else:
                    left = mid + 1
            else:
                if target > nums[mid] and target <= nums[right]:
                    left = mid + 1
                else:
                    right = mid - 1
                    
        return -1
            
        

File: 33-search-in-rotated-sorted-array/33-search-in-rotated-sorted-array.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number}
 */
var search = function(nums, target) {
    let l = 0
    let r = nums.length - 1
    while ( l <= r) {
        const mid = Math.floor((l+r)/2)
        if (target === nums[mid]) {
            return mid
        }
        
        // separate cases: what portion of the array are we in, left or right
        // left portion
        if (nums[mid] >= nums[l]) {
            if (target > nums[mid] || target < nums[l]) {
                l = mid + 1
            } else {
                r = mid - 1
            }
        } else {
            if (target < nums[mid] || target > nums[r]) {
                r = mid - 1
            } else {
                l = mid + 1
            }
        }
    }
    
    return -1
};